Developing-new-features-with-feature-gating.md

```diff
@@ -41,6 +44,40 @@ export enum FeatureNames {
 }
 ```
 
+### Modifying unit tests to account for the creation of the new flag
```

Comment:
I'm wondering a couple of things:

- What is the desired practice for unit tests -- should they test both the case when the flag is off and when the flag is on? If so, maybe add instructions on how to turn a flag on in the tests.
- If this section isn't for the previous point, then what is it for exactly? I'm not quite sure; more explanation might be helpful.

Reply:
The unit tests aren't about the flags being on/off. Basically, there are two specific tests (one in the backend and one in the frontend) whose success or failure depends on them being aware of all the feature flags that currently exist.

So when a new feature flag is registered in the codebase by a developer, they need to make these unit tests aware of that, otherwise these two tests will fail. Hence it is a part of the feature-flag creation process.

Does that answer the question?

Reply:
Yup (though maybe you could explain this a bit more in the description of these changes).

But, in addition to the above, do you want to explain to developers how to test stuff that is gated by a feature flag? This should probably be part of "standard procedure", too.

Reply:
I've added instructions for unit testing gated features in the frontend.

However, I'm not very well-versed in Oppia's backend, nor do I have much experience writing unit tests in it. Have any of the other contributors worked with the gating system on the backend? Meanwhile, I'll try looking into `platform_feature_services_test.py` to see if any unit tests there might work.

Reply:
This I don't really know, sorry -- though @BenHenning might since he originally oversaw the project. I think looking at the file you mentioned is a good first step, FWIW.

@BenHenning if you have any comments on this, could you please leave them here? Thanks!

Reply:
I would've suggested the same thing (to look at existing cases of platform parameters being used in tests both in Oppia web and backend). I'm not sure there's better references or documentation available right now other than the actual code, however.

Regarding the actual best practice here, I think we should encourage testing functionality both with the flag on and off (since both states will exist for users at some point in time, so we need to make sure that the correct thing happens in both case rather than relying on the default flag state).

Reply:
> we should encourage testing functionality both with the flag on and off (since both states will exist for users at some point in time, so we need to make sure that the correct thing happens in both case rather than relying on the default flag state)

Yep, that's actually how the frontend tests are currently written (mock `PlatformFeatureService` and test with the feature status getter set to return both `false` and `true` in two separate tests). Makes sense to test the backend side of things in a similar manner.